### PR TITLE
Fix TimersTest

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

Fix `TimersTest.clearInterval`

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/web/TimersTest.java [hunk 1]
```diff
@@ -119,7 +119,7 @@ public void clearInterval() throws Exception {
         var p = function (x, y) {
             globalThis.result += 10*x + y;
         };
-        var intervalId = setInterval(p, 10, 4, 2);
+        var intervalId = setInterval(p, 99999, 4, 2);
```
Comment (4e6) [inline lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/web/TimersTest.java:122]: I'm surprised it worked at all. 10ms is a tight gap

Comment (Frizi) [reply lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/web/TimersTest.java:122]: In most JS environments, even 0 here would be guaranteed to not actually run the callback if you clear it synchronously within the same task. Also 10ms sounds like a lot for most things, that's more than 1 frame budget for 120hz rendering.

**Review by Frizi — APPROVED** [review]
This fix is important to unblock CI, but I'd appreciate if the implementation was actually supporting the test scenario as it was.

File: lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/web/TimersTest.java (hunk 1 above)
Comment (JaroslavTulach) [reply lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/web/TimersTest.java:122]: > even 0 here would be guaranteed to not actually run the callback if you clear it synchronously within the same task

In such case we probably need to fix the task scheduler and not to change the test...

**Comment by JaroslavTulach** [comment]
We should simplify the code first. What about having just a single executor instead of two:
```diff
diff --git lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java
index ff6d8ece8f..f817728456 100644
--- lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java
+++ lib/java/ydoc-server/src/main/java/org/enso/ydoc/Main.java
@@ -25,7 +25,7 @@ public class Main {
     Sampling.init();
 
     // Can't use try-with-resource in ExecutorService because API was added in JDK19
-    var executor = Executors.newSingleThreadExecutor();
+    var executor = Executors.newSingleThreadScheduledExecutor();
     try {
       try (var parser = new ParserPolyfill()) {
         var ydocJs =
diff --git lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java
index 84d0a0cc71..059ac0d7af 100644
--- lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java
+++ lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java
@@ -1,7 +1,5 @@
 package org.enso.ydoc.polyfill.web;
 
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
@@ -25,28 +23,21 @@ final class Timers extends PolyfillBase implements ProxyExecutable {
   private static final String TIMERS_JS = "timers.js";
 
   private static final TimeUnit TIME_UNIT = TimeUnit.MILLISECONDS;
+  private final ScheduledExecutorService executor;
 
-  private final ScheduledExecutorService scheduledExecutor =
-      Executors.newSingleThreadScheduledExecutor(
-          (r) -> {
-            var thread = new Thread(r);
-            thread.setDaemon(true);
-            return thread;
-          });
-
-  private final ExecutorService executor;
-
-  Timers(ExecutorService executor) {
+  Timers(ScheduledExecutorService executor) {
     super(TIMERS_JS);
     this.executor = executor;
   }
 
-  public Object setTimeout(Value func, long delay, Value[] args) {
-    return scheduledExecutor.schedule(execute(func, args), delay, TIME_UNIT);
+  public Future<?> setTimeout(Value func, long delay, Value[] args) {
+    Runnable action = () -> func.executeVoid(args);
+    return executor.schedule(action, delay, TIME_UNIT);
   }
 
-  public Object setInterval(Value func, long delay, Value[] args) {
-    return scheduledExecutor.scheduleAtFixedRate(execute(func, args), delay, delay, TIME_UNIT);
+  public Future<?> setInterval(Value func, long delay, Value[] args) {
+    Runnable action = () -> func.executeVoid(args);
+    return executor.scheduleAtFixedRate(action, delay, delay, TIME_UNIT);
   }
 
   public void clearTimeout(Object actionId) {
@@ -91,8 +82,4 @@ final class Timers extends PolyfillBase implements ProxyExecutable {
       default -> throw new IllegalStateException(command);
     };
   }
-
-  private Runnable execute(Value func, Object[] args) {
-    return () -> executor.execute(() -> func.executeVoid(args));
-  }
 }
diff --git lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/WebEnvironment.java lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/WebEnvironment.java
index 317a65888a..db8f1fb792 100644
--- lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/WebEnvironment.java
+++ lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/WebEnvironment.java
@@ -1,6 +1,6 @@
 package org.enso.ydoc.polyfill.web;
 
-import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
 import org.graalvm.polyglot.Context;
 import org.graalvm.polyglot.HostAccess;
 
@@ -12,7 +12,7 @@ public final class WebEnvironment {
 
   private WebEnvironment() {}
 
-  public static void initialize(Context ctx, ExecutorService executor) {
+  public static void initialize(Context ctx, ScheduledExecutorService executor) {
     var performance = new Performance();
     performance.initialize(ctx);
 
diff --git lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/ExecutorSetup.java lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/ExecutorSetup.java
index 6ad6b46abd..25c95b1b4b 100644
--- lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/ExecutorSetup.java
+++ lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/ExecutorSetup.java
@@ -2,19 +2,19 @@ package org.enso.ydoc.polyfill;
 
 import static org.junit.Assert.fail;
 
-import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import org.junit.After;
 import org.junit.Before;
 
 public abstract class ExecutorSetup {
 
-  protected ExecutorService executor;
+  protected ScheduledExecutorService executor;
 
   @Before
   public void setup() throws Exception {
-    executor = Executors.newSingleThreadExecutor();
+    executor = Executors.newSingleThreadScheduledExecutor();
   }
 
   @After
```

**Review by JaroslavTulach — COMMENTED** [review]
The two executors in `Timers` are _complicated_ and buggy as they can _run in parallel_ causing the test to fail.

> even 0 here would be guaranteed to not actually run the callback if you clear it synchronously within the same task

That can only be guaranteed with a single threaded executor - and that's the reason why we **need only one executor**.

File: lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/web/TimersTest.java (hunk 1 above)
Comment (JaroslavTulach) [reply lib/java/ydoc-server/src/test/java/org/enso/ydoc/polyfill/web/TimersTest.java:122]: The merger was faster than me...

@4e6 please [fix the problem again](https://github.com/enso-org/enso/pull/10118#issuecomment-2136490798) and `setInterval` to `1` or `0` to verify the fix is correct.

**Pull request merged**